File: ios/CodexMeterCore/Sources/CodexMeterCore/AdaptiveRefreshPolicy.swift
import Foundation

/// Pure, low-cost policy for selecting the next automatic usage refresh.
public enum AdaptiveRefreshPolicy {
    private static let intervals = [5, 10, 15, 30, 60, 120]

    public static func chooseMinutes(
        snapshot: UsageSnapshot?,
        attentionScore: Double,
        localHour: Int,
        consecutiveFailures: Int,
        now: Date = Date()
    ) -> Int {
        let windows = [snapshot?.fiveHour, snapshot?.weekly, snapshot?.monthly]
            .compactMap { $0 }
            .filter {
                guard let reset = $0.effectiveResetDate(relativeTo: snapshot?.fetchedAt ?? now) else {
                    return true
                }
                return reset > now
            }
        let remaining = windows.map(\.remainingPercent).min()
        let limited = snapshot.map { !$0.allowed || $0.limitReached } ?? false

        var minutes = 30
        if let remaining {
            switch remaining {
            case ...10:
                minutes = 5
            case ...25:
                minutes = 10
            case ...50:
                minutes = 15
            case ...75:
                minutes = 30
            default:
                minutes = 60
            }
        }
        if limited {
            minutes = 5
        }

        let nextUsedReset = windows
            .filter { $0.usedPercent > 0 }
            .compactMap { $0.effectiveResetDate(relativeTo: snapshot?.fetchedAt ?? now) }
            .filter { $0 > now }
            .min()
        let untilReset = nextUsedReset?.timeIntervalSince(now) ?? .infinity
        if untilReset <= 15 * 60 {
            minutes = min(minutes, 5)
        } else if untilReset <= 60 * 60 {
            minutes = min(minutes, 10)
        }

        if UsagePace.mostAcceleratedWindow(in: snapshot, now: now) != nil {
            minutes = min(minutes, 10)
        }

        let attention = max(0, attentionScore.isFinite ? attentionScore : 0)
        switch attention {
        case 6...:
            minutes = min(minutes, 5)
        case 3...:
            minutes = min(minutes, 10)
        case 1.5...:
            minutes = min(minutes, 15)
        case 0.5...:
            minutes = min(minutes, 30)
        default:
            break
        }

        let urgent = limited || (remaining ?? 101) <= 25 || untilReset <= 60 * 60
        let hour = min(23, max(0, localHour))
        if !urgent, attention < 0.5, (1..<6).contains(hour) {
            minutes = max(minutes, 120)
        }

        for _ in 0..<min(3, max(0, consecutiveFailures)) {
            minutes = nextSlower(than: minutes)
        }
        return urgent ? min(minutes, 30) : min(minutes, 120)
    }

    private static func nextSlower(than minutes: Int) -> Int {
        intervals.first(where: { $0 > minutes }) ?? intervals[intervals.count - 1]
    }
}

File: android/shared/src/main/java/dev/scorpion7slayer/modelsmeter/AdaptiveRefreshPolicy.java
package dev.scorpion7slayer.modelsmeter;

import java.util.concurrent.TimeUnit;

/** Pure, low-cost policy for selecting the next automatic usage refresh. */
public final class AdaptiveRefreshPolicy {
    private static final int[] INTERVALS = {5, 10, 15, 30, 60, 120};

    private AdaptiveRefreshPolicy() {
    }

    public static int chooseMinutes(UsageSnapshot snapshot, double attentionScore,
            int localHour, int consecutiveFailures, long nowMillis) {
        UsageWindow fiveHour = snapshot == null ? null
                : UsageSnapshot.currentWindow(snapshot.fiveHour, snapshot.fetchedAtMillis,
                        nowMillis);
        UsageWindow weekly = snapshot == null ? null
                : UsageSnapshot.currentWindow(snapshot.weekly, snapshot.fetchedAtMillis,
                        nowMillis);
        UsageWindow monthly = snapshot == null ? null
                : UsageSnapshot.currentWindow(snapshot.monthly, snapshot.fetchedAtMillis,
                        nowMillis);
        int remaining = minimumRemaining(fiveHour, weekly, monthly);
        boolean limited = snapshot != null && (!snapshot.allowed || snapshot.limitReached);

        int minutes;
        if (limited || remaining <= 10) {
            minutes = 5;
        } else if (remaining <= 25) {
            minutes = 10;
        } else if (remaining <= 50) {
            minutes = 15;
        } else if (remaining <= 75) {
            minutes = 30;
        } else if (remaining <= 100) {
            minutes = 60;
        } else {
            minutes = 30;
        }

        long nextUsedReset = nextUsedReset(
                snapshot == null ? 0L : snapshot.fetchedAtMillis, nowMillis,
                fiveHour, weekly, monthly);
        long untilReset = nextUsedReset <= nowMillis ? Long.MAX_VALUE : nextUsedReset - nowMillis;
        if (untilReset <= TimeUnit.MINUTES.toMillis(15)) {
            minutes = Math.min(minutes, 5);
        } else if (untilReset <= TimeUnit.HOURS.toMillis(1)) {
            minutes = Math.min(minutes, 10);
        }

        if (snapshot != null
                && UsagePace.mostAcceleratedWindow(snapshot, nowMillis, UsagePace.BALANCED)
                != UsagePace.WINDOW_NONE) {
            minutes = Math.min(minutes, 10);
        }

        double attention = Math.max(0.0d, attentionScore);
        if (attention >= 6.0d) {
            minutes = Math.min(minutes, 5);
        } else if (attention >= 3.0d) {
            minutes = Math.min(minutes, 10);
        } else if (attention >= 1.5d) {
            minutes = Math.min(minutes, 15);
        } else if (attention >= 0.5d) {
            minutes = Math.min(minutes, 30);
        }

        boolean urgent = limited || remaining <= 25
                || untilReset <= TimeUnit.HOURS.toMillis(1);
        int hour = Math.max(0, Math.min(23, localHour));
        if (!urgent && attention < 0.5d && hour >= 1 && hour < 6) {
            minutes = Math.max(minutes, 120);
        }

        int failures = Math.max(0, Math.min(3, consecutiveFailures));
        for (int i = 0; i < failures; i++) {
            minutes = nextSlower(minutes);
        }
        return urgent ? Math.min(minutes, 30) : Math.min(minutes, 120);
    }

    private static int minimumRemaining(UsageWindow... windows) {
        int remaining = 101;
        for (UsageWindow window : windows) {
            if (window != null) remaining = Math.min(remaining, window.remainingPercent());
        }
        return remaining;
    }

    private static long nextUsedReset(long observedAtMillis, long nowMillis,
            UsageWindow... windows) {
        long next = Long.MAX_VALUE;
        for (UsageWindow window : windows) {
            if (window != null && window.usedPercent > 0) {
                long reset = window.effectiveResetAtMillis(observedAtMillis);
                if (reset > nowMillis) next = Math.min(next, reset);
            }
        }
        return next == Long.MAX_VALUE ? 0L : next;
    }

    private static int nextSlower(int minutes) {
        for (int interval : INTERVALS) {
            if (interval > minutes) return interval;
        }
        return INTERVALS[INTERVALS.length - 1];
    }
}
